### PR TITLE
Only call into reducer if torch.is_grad_enabled()

### PR DESCRIPTION
**Pull request by pietern**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#19901 Finer grained consistency check in reducer**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15118871/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19897 Only call into reducer if torch.is_grad_enabled()&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D15118726/)

During validation, gradient reduction is not needed, and autograd is
never called. The model output will always be a detached tensor. After
the new reducer was merged, this meant that it would find all model
parameters unused, and kick off reduction for them. When #19799 and
#19821 were merged it looked like model output during validation is an
output where no parameters are used and it tries to kick off reduction
of zeroed gradients. Test for `torch.is_grad_enabled()` and
`self.training` before calling into the reducer.

Differential Revision: [D15118726](https://our.internmc.facebook.com/intern/diff/D15118726/)

### CODE REVIEW

**Review by mrshenli — APPROVED** [review]

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -380,14 +380,16 @@ def forward(self, *inputs, **kwargs):
...
 
-        # We'll return the output object verbatim since it is a freeform object.
-        # We need to find any tensors in this object, though, because we need to
-        # figure out which parameters were used during this forward pass,
-        # to ensure we short circuit reduction for any unused parameters.
-        if self.find_unused_parameters:
-            self.reducer.prepare_for_backward(list(_find_tensors(output)))
-        else:
-            self.reducer.prepare_for_backward([])
+        if self.training and torch.is_grad_enabled():
```
Comment (soumith) [inline torch/nn/parallel/distributed.py]: self.training has nothing to do with whether gradients are called or not. It's only used for whether BatchNorm and Dropout are in training or eval mode.
Dont check for this as it isn't hard indicator of training or eval mode (only a hint)

**Review by soumith — CHANGES_REQUESTED** [review]

File: torch/nn/parallel/distributed.py (hunk 1 above)
Comment (pietern) [reply torch/nn/parallel/distributed.py]: Removed the `self.training` check.

**Review by soumith — APPROVED** [review]
